swift/ql/lib/codeql/swift/elements/expr/DeclRefExprConstructor.qll

```diff
@@ -1,7 +1,8 @@
 private import codeql.swift.generated.Raw
+private import codeql.swift.elements.expr.MethodLookupExprConstructor
 
 predicate constructDeclRefExpr(Raw::DeclRefExpr id) {
-  // exclude an argument that will be part of a DotSyntaxCallExpr
-  // that will be transformed into a MethodRefCallExpr
-  not exists(Raw::DotSyntaxCallExpr c | id = c.getFunction())
+  // exclude an argument that will be part of a SelfApplyExpr
+  // that will be transformed into a MethodLookupExpr
+  not exists(Raw::SelfApplyExpr e | id.getDecl() = extractDeclFromSelfApplyExpr(e))
```

Comment:
hmm, I _think_ this is wrong for two reasons.

First and foremost, I don't think we should be removing any `DeclRefExpr` at all. If you look at [what we discussed before](https://github.com/github/codeql-c-team/issues/1362#issuecomment-1314955181), the idea was to preserve `DeclRefExpr` underneath the new `MethodLookupExpr`, and not hide it any more. The downside of doing what is currently done here is that even though we have not seen it yet, I don't see anything preventing a `SelfApplyExpr` to not have a `DeclRefExpr` underneath it, so while we would still synthesize a `MethodLookupExpr` out of it, it would be incomplete. Maximum genericity is only attainable by keeping the `method` part of the `MethodLookupExpr` a generic `Expr`, not a `Decl`. The other problem, is that this leaves all possible `ImplicitConversionExpr` going from `SelfApplyExpr` down to `DeclRefExpr` as orphans in the AST tree. To keep this up, we should strip those out as well, but it becomes complex as it requires changing the constructors of all the concrete subclasses of `ImplicitConversionExpr`, which becomes a bit of a nightmare :sweat_smile: Sorry I really did not notice this earlier :bow: 

Even disregarding the above, this will remove all `DeclRefExpr` instances referring to a decl that has any even unrelated `DeclRefExpr` packed inside a `SelfApplyExpr`. Even ignoring what I wrote above, we would still want to instead only remove those specific `DeclRefExpr` packed inside a `SelfApplyExpr`.